File: Beakjoon/Cpp/1504 특정한 최단 경로.cpp
// 반드시 지나야 하는 정점에 대해 역방향, 정방향 모두 검사를 해야함.
// 놓친 부분은 양방향그래프인것을 못봤음..

#include <bits/stdc++.h>
#define INF 9999999

using namespace std;

int n, m;
vector<pair<int, int>> graph[801];

int dijkstra(int start, int end) {
    int dist[801] = {0};
    for(int i = 0 ; i < 801; i++) {
        dist[i] = INF;
    }
    
    dist[start] = 0;
    priority_queue<pair<int, int>> pq;
    pq.push(make_pair(0, start));
    
    while(!pq.empty()) {
        int now_dist = -pq.top().first; int now_node = pq.top().second;
        pq.pop();
        
        for(auto x : graph[now_node]) {
            
            int next_dist = x.first; int next_node = x.second;
            if(now_dist + next_dist < dist[next_node]) {
                dist[next_node] = now_dist + next_dist;
                pq.push(make_pair(-dist[next_node], next_node));
            }
        }
    }
    
    return dist[end];
}

int main() {
    cin >> n >> m;
    for(int i = 0 ; i < m; i++) {
        int a,b,c; cin >> a >> b >> c;
        graph[a].push_back(make_pair(c,b));
        graph[b].push_back(make_pair(c,a));
    }
    int v1, v2; cin >> v1 >> v2;
    
    int answer = min(dijkstra(1, v2) + dijkstra(v2, v1) + dijkstra(v1, n),
                dijkstra(1, v1) + dijkstra(v1, v2) + dijkstra(v2, n));
    
    if(answer < 0 || answer >= INF) cout << -1;
    else cout << answer;
}

File: Beakjoon/Javascript/1504. 특정한 최단 경로.js
class Node {
  constructor(val, priority) {
    this.val = val;
    this.priority = priority;
  }
}

class PriorityQueue {
  constructor() {
    this.values = [];
  }

  enqueue(val, priority) {
    let newNode = new Node(val, priority);
    this.values.push(newNode);
    this.bubbleUp();
  }

  dequeue() {
    const min = this.values[0];
    const end = this.values.pop();
    if (this.values.length > 0) {
      this.values[0] = end;
      this.bubbleDown();
    }
    return min;
  }

  bubbleUp() {
    let idx = this.values.length - 1;
    const element = this.values[idx];
    while (idx > 0) {
      let parentIdx = Math.floor((idx - 1) / 2);
      let parent = this.values[parentIdx];
      if (element.priority >= parent.priority) break;
      this.values[parentIdx] = element;
      this.values[idx] = parent;
      idx = parentIdx;
    }
  }

  bubbleDown() {
    let idx = 0;
    const length = this.values.length;
    const element = this.values[0];
    while (true) {
      let leftChildIdx = 2 * idx + 1;
      let rightChildIdx = 2 * idx + 2;
      let leftChild, rightChild;
      let swap = null;

      if (leftChildIdx < length) {
        leftChild = this.values[leftChildIdx];
        if (leftChild.priority < element.priority) {
          swap = leftChildIdx;
        }
      }
      if (rightChildIdx < length) {
        rightChild = this.values[rightChildIdx];
        if (
          (swap === null && rightChild.priority < element.priority) ||
          (swap !== null && rightChild.priority < leftChild.priority)
        ) {
          swap = rightChildIdx;
        }
      }

      if (swap === null) break;
      this.values[idx] = this.values[swap];
      this.values[swap] = element;
      idx = swap;
    }
  }
}

const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");
const arr = input.map((el) => el.split(" ").map((el) => +el));
const [n, e] = arr.shift();
const [v1, v2] = arr.pop();
const graph = {};
let answer = 0;

for (let i = 1; i <= n; i++) {
  graph[i] = [];
}

for (let i = 0; i < e; i++) {
  const [s, e, w] = arr[i];
  graph[s].push([e, w]);
  graph[e].push([s, w]);
}

const dijkstra = (start, end) => {
  const distances = Array.from({ length: n + 1 }, () => 0);
  const initDistance = () => {
    for (let i = 0; i <= n; i++) {
      distances[i] = Infinity;
    }
  };

  initDistance();

  const q = new PriorityQueue();
  distances[start] = 0;
  q.enqueue([start, 0], 0);

  while (q.values.length) {
    const [nowNode, nowDist] = q.dequeue().val;

    if (nowNode === end) break;

    for (let x of graph[nowNode]) {
      const [nextNode, nextDist] = x;
      if (nowDist + nextDist < distances[nextNode]) {
        distances[nextNode] = nowDist + nextDist;
        q.enqueue([nextNode, distances[nextNode]], distances[nextNode]);
      }
    }
  }
  return distances[end];
};

answer = Math.min(
  dijkstra(1, v1) + dijkstra(v1, v2) + dijkstra(v2, n),
  dijkstra(1, v2) + dijkstra(v2, v1) + dijkstra(v1, n)
);

console.log(answer === Infinity ? -1 : answer);
